common/persistence/sql/sqlplugin/auth/rds_auth_plugin.go

```diff
@@ -0,0 +1,150 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package auth
+
+import (
+	"context"
+	"encoding/base64"
+	"io/ioutil"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	AWSConfig "github.com/aws/aws-sdk-go-v2/config"
+	AWSAuth "github.com/aws/aws-sdk-go-v2/feature/rds/auth"
+
+	"go.temporal.io/server/common/auth"
+	"go.temporal.io/server/common/config"
+)
+
+const defaultTimeout = time.Second * 10
+const rdsCaUrl = "https://s3.amazonaws.com/rds-downloads/rds-combined-ca-bundle.pem"
+
+var rdsAuthFn = AWSAuth.BuildAuthToken
+
+func init() {
+	RegisterPlugin("rds-iam-auth", NewRDSAuthPlugin(nil))
+}
+
+func fetchRdsCA(ctx context.Context) (string, error) {
+	ctx, cancel := context.WithTimeout(ctx, defaultTimeout)
+	defer cancel()
+
+	req, err := http.NewRequestWithContext(ctx, "GET", rdsCaUrl, nil)
+	if err != nil {
+		return "", err
+	}
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return "", err
+	}
+
+	defer resp.Body.Close()
+	pem, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+
+	return base64.StdEncoding.EncodeToString(pem), nil
+}
+
+type RDSAuthPlugin struct {
+	awsConfig        *aws.Config
+	rdsPemBundle     string
+	initRdsPemBundle sync.Once
+}
+
+func NewRDSAuthPlugin(awsConfig *aws.Config) AuthPlugin {
+	return &RDSAuthPlugin{
+		awsConfig: awsConfig,
+	}
+}
+
+func (plugin *RDSAuthPlugin) getToken(ctx context.Context, addr string, region string, user string, credentials aws.CredentialsProvider) (string, error) {
+	reqCtx, cancel := context.WithTimeout(ctx, defaultTimeout)
+	defer cancel()
+
+	return rdsAuthFn(reqCtx, addr, region, user, credentials)
+}
+
+func (plugin *RDSAuthPlugin) resolveAwsConfig(ctx context.Context) (*aws.Config, error) {
+	if plugin.awsConfig != nil {
+		return plugin.awsConfig, nil
+	}
+
+	reqCtx, cancel := context.WithTimeout(ctx, defaultTimeout)
+	defer cancel()
+
+	cfg, err := AWSConfig.LoadDefaultConfig(reqCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &cfg, nil
+}
+
+func (plugin *RDSAuthPlugin) GetConfig(ctx context.Context, cfg *config.SQL) (*config.SQL, error) {
+	awsCfg, err := plugin.resolveAwsConfig(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	token, err := plugin.getToken(ctx, cfg.ConnectAddr, awsCfg.Region, cfg.User, awsCfg.Credentials)
+	if err != nil {
+		return nil, err
+	}
+
+	cfg.Password = token
+	cfg.ConnectProtocol = "tcp"
+
+	if cfg.ConnectAttributes == nil {
+		cfg.ConnectAttributes = map[string]string{}
+	}
+
+	// mysql requires this plugin to use the token as a password
+	if cfg.PluginName == "mysql" {
+		cfg.ConnectAttributes["allowCleartextPasswords"] = "true"
+	}
+
+	// if TLS is not configured, we default to the RDS CA
+	// this is required for mysql to send cleartext passwords
+	if cfg.TLS == nil {
+		plugin.initRdsPemBundle.Do(func() {
+			ca, err := fetchRdsCA(ctx)
+			if err != nil {
+				return
+			}
```

Comment:
hmmm, I'm sorry I didn't think about this error case. But I think we need to handle this error. When this failed, we should fail this whole GetConfig() call.

Reply:
No worries, I added handling for the 2 error cases I could foresee. Let me know if you see any others.